packages/git-diff/lib/git-diff-view.js

```diff
@@ -72,122 +151,106 @@ module.exports = class GitDiffView {
   }
 
-  moveToLineNumber(lineNumber) {
-    if (lineNumber != null && lineNumber >= 0) {
-      this.editor.setCursorBufferPosition([lineNumber, 0]);
-      this.editor.moveToFirstCharacterOfLine();
+  updateIconDecoration() {
+    const editorElement = atom.views.getView(this.editor);
+    const gutter = editorElement.querySelector('.gutter');
+    if (gutter) {
+      if (
+        atom.config.get('editor.showLineNumbers') &&
+        atom.config.get('git-diff.showIconsInEditorGutter')
+      ) {
+        gutter.classList.add('git-diff-icon');
+      } else {
+        gutter.classList.remove('git-diff-icon');
+      }
     }
   }
 
-  subscribeToRepository() {
-    this.repository = repositoryForPath(this.editor.getPath());
-    if (this.repository) {
-      this.subscriptions.add(
-        this.repository.onDidChangeStatuses(() => {
-          this.scheduleUpdate();
-        })
-      );
-      this.subscriptions.add(
-        this.repository.onDidChangeStatus(changedPath => {
-          if (changedPath === this.editor.getPath()) this.scheduleUpdate();
-        })
-      );
+  moveToLineNumber(lineNumber) {
+    if (lineNumber != null) {
+      this.editor.setCursorBufferPosition([lineNumber, 0]);
+      this.editor.moveToFirstCharacterOfLine();
     }
   }
 
-  cancelUpdate() {
-    clearImmediate(this.immediateId);
-  }
-
   scheduleUpdate() {
-    this.cancelUpdate();
-    this.immediateId = setImmediate(this.updateDiffs);
+    // Use Chromium native requestAnimationFrame because it yields
+    // to the browser, is standard and doesn't involve extra JS overhead.
+    if (this._animationId) cancelAnimationFrame(this._animationId);
+    this._animationId = requestAnimationFrame(this.updateDiffs);
   }
 
+  /**
+   * @describe Uses text markers in the target editor to visualize
+   *   git modifications, additions, and deletions. The current algorithm
+   *   just redraws the markers each call.
+   */
   updateDiffs() {
-    if (this.editor.isDestroyed()) return;
-    this.removeDecorations();
-    const path = this.editor && this.editor.getPath();
-    if (
-      path &&
-      this.editor.getBuffer().getLength() < MAX_BUFFER_LENGTH_TO_DIFF
-    ) {
-      this.diffs =
-        this.repository &&
-        this.repository.getLineDiffs(path, this.editor.getText());
-      if (this.diffs) this.addDecorations(this.diffs);
-    }
-  }
+    const bufferLength = this.editor.getBuffer().getLength();
+    if (bufferLength < MAX_BUFFER_LENGTH_TO_DIFF) {
+      const path = this.editor.getPath();
+
+      // Before we redraw the diffs, tear down the old markers.
+      if (this.diffs)
+        for (const diff of this.diffs) this.markers.get(diff).destroy();
+
+      // WARNING: Could cause future memory leak if git-utils ever
+      // changes their diff strategy to one that re-uses diffs between
+      // requests. But that's unlikely to happen.
+      this.diffs = this.repository.getLineDiffs(path, this.editor.getText());
```

Comment:
You called `this.editor.getBuffer` in the above. You can get the text directly from the buffer. Please cache the buffer and call the needed methods.